datashader/datatypes.py

```diff
@@ -0,0 +1,364 @@
+import numpy as np
+from pandas.api.extensions import (
+    ExtensionDtype, ExtensionArray, register_extension_dtype)
+from numbers import Integral
+
+
+@register_extension_dtype
+class RaggedDtype(ExtensionDtype):
+    name = 'ragged'
+    type = np.ndarray
+    base = np.dtype('O')
+
+    @classmethod
+    def construct_array_type(cls):
+        return RaggedArray
+
+    @classmethod
+    def construct_from_string(cls, string):
+        if string == 'ragged':
+            return RaggedDtype()
+        else:
+            raise TypeError("Cannot construct a '{}' from '{}'"
+                            .format(cls, string))
+
+
+class RaggedArray(ExtensionArray):
+    def __init__(self, data, dtype=None):
+        """
+        Construct a RaggedArray
+
+        Parameters
+        ----------
+        data
+            List or numpy array of lists or numpy arrays
```

Comment:
Seems that a `dict` of the correct schema is also accepted. Or is that intended to be "private"? Could that be placed in an alternative constructor `_from_dict`?

It may also be useful to accept a `RaggedArray` here..

Reply:
I was initially thinking that the `dict` form would be private so that there would be an internal fast-path for constructing RaggedArrays without going through the nested array representation.  But I went ahead and made the `mask`, `flat_array`, and `start_indices` properties public so I might as well document the `dict` constructor form so that other libraries can use the fast-path as well.